compiler/optimizer/ValuePropagationCommon.cpp

```diff
@@ -2429,25 +2463,47 @@ void OMR::ValuePropagation::generateArrayTranslateNode(TR::TreeTop *callTree,TR:
-      node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, dstOff, hdrSize);
-      dst = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, dstObj, node);
-      arrayTranslateNode->setSourceIsByteArrayTranslate(false);
-      arrayTranslateNode->setTargetIsByteArrayTranslate(true);
+      if (encode)
+         {
+         srcOff = TR::TransformUtil::generateArrayOffsetTrees(comp(), srcOff, strideNode, 0, false);
+         arrayTranslateNode->setSourceIsByteArrayTranslate(false);
+         arrayTranslateNode->setTargetIsByteArrayTranslate(true);
+         }
+      else
+         {
+         dstOff = TR::TransformUtil::generateArrayOffsetTrees(comp(), dstOff, strideNode, 0, false);
+         arrayTranslateNode->setSourceIsByteArrayTranslate(true);
+         arrayTranslateNode->setTargetIsByteArrayTranslate(false);
+         }
+      src = TR::TransformUtil::generateArrayAddressTrees(comp(), srcObj, srcOff);
+      dst = TR::TransformUtil::generateArrayAddressTrees(comp(), dstObj, dstOff);
       }
    else
+#endif /* OMR_GC_SPARSE_HEAP_ALLOCATION */
       {
-      node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, srcOff, hdrSize);
-      src = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, srcObj, node);
-      node = TR::Node::create(is64BitTarget ? TR::lmul : TR::imul, 2, dstOff, strideNode);
-      node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, node, hdrSize);
-      dst = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, dstObj, node);
-      arrayTranslateNode->setSourceIsByteArrayTranslate(true);
-      arrayTranslateNode->setTargetIsByteArrayTranslate(false);
+      if (encode)
+         {
+         node = TR::Node::create(is64BitTarget ? TR::lmul : TR::imul, 2, srcOff, strideNode);
+         node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, node, hdrSize);
+         src = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, srcObj, node);
+         node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, dstOff, hdrSize);
+         dst = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, dstObj, node);
+         arrayTranslateNode->setSourceIsByteArrayTranslate(false);
+         arrayTranslateNode->setTargetIsByteArrayTranslate(true);
+         }
+      else
+         {
+         node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, srcOff, hdrSize);
+         src = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, srcObj, node);
+         node = TR::Node::create(is64BitTarget ? TR::lmul : TR::imul, 2, dstOff, strideNode);
+         node = TR::Node::create(is64BitTarget ? TR::ladd : TR::iadd, 2, node, hdrSize);
+         dst = TR::Node::create(is64BitTarget? TR::aladd : TR::aiadd, 2, dstObj, node);
+         arrayTranslateNode->setSourceIsByteArrayTranslate(true);
+         arrayTranslateNode->setTargetIsByteArrayTranslate(false);
+         }
```

Comment:
Was this second path added back only because of failures with arraylets? If so, maybe we should add a TODO with some context of the issue, and/or link to an issue if one exists?

Reply:
Thumbs down on the TODO, but thumbs up for creating and referencing an issue.

Reply:
The added path is for offheap while keeping the previous path for non-offheap cases (just indentation change). The possible improvement is merging the paths to work on the all cases, I can open an issue for that.

Reply:
Created an issue, and as explained and given that there's multiple spots that can be improved to consolidated paths I didn't specifically add a comment in the code about it here.
https://github.com/eclipse/omr/issues/7333